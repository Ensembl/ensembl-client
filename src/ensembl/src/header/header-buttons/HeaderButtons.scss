.button {
  display: inline-block;
  bottom: 1px;
  height: 17px;
  margin-left: 20px;
  position: relative;
  width: 17px;
  vertical-align: middle;
}

.headerButtonDisabled {
  svg {
    fill: #7d8e9b;
  }
}
