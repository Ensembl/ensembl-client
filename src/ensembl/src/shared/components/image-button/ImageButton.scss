@import 'src/styles/common';

.imageButton {
  position: relative;
  font-size: 0;

  svg {
    width: 100%;
    height: 100%;
    transition: opacity 0.4s;
  }

  button {
    width: 100%;
    height: 100%;
  }
}

.default {
  svg {
    fill: $blue;
  }
}

.selected {
  svg {
    fill: $white;
    background: $blue;
  }
  &svg:hover {
    opacity: 0.8;
  }
}

.unselected {
  svg {
    fill: $blue;
  }
  &svg:hover {
    opacity: 0.8;
  }
}

.disabled {
  svg {
    fill: $grey;
    opacity: 1;
  }
  button {
    cursor: default;
  }
}
