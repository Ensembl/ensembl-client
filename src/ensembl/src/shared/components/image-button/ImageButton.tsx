import React from 'react';

import useHover from 'src/shared/hooks/useHover';

import defaultStyles from './ImageButton.scss';
import ImageHolder from './ImageHolder';
import Tooltip from 'src/shared/components/tooltip/Tooltip';

import imageButtonStyles from './ImageButton.scss';

import { Status } from 'src/shared/types/status';

type ImageButtonStatus =
  | Status.ACTIVE
  | Status.INACTIVE
  | Status.DISABLED
  | Status.DEFAULT
  | Status.HIGHLIGHTED;

type Props = {
  buttonStatus: ImageButtonStatus;
  description: string;
  image: React.FunctionComponent<React.SVGProps<SVGSVGElement>> | string;
  classNames?: { [key in ImageButtonStatus]?: string };
  onClick?: () => void;
};

const ImageButton = (props: Props) => {
  const [hoverRef, isHovered] = useHover<HTMLDivElement>();

  const handleClick = () => {
    props.onClick && props.onClick();
  };

  const buttonProps =
    props.buttonStatus === Status.DISABLED ? {} : { onClick: handleClick };

  const { classNames, ...rest } = props;

  const styles = classNames
    ? { ...defaultStyles, ...props.classNames }
    : defaultStyles;

  const shouldShowTooltip = Boolean(props.description) && isHovered;

  return (
    <div
      ref={hoverRef}
      className={imageButtonStyles.imageButton}
      {...buttonProps}
    >
      <ImageHolder {...rest} classNames={styles} />
      {shouldShowTooltip && (
        <Tooltip autoAdjust={true}>{props.description}</Tooltip>
      )}
    </div>
  );
};

ImageButton.defaultProps = {
  buttonStatus: Status.DEFAULT,
  description: '',
  image: ''
};

export default ImageButton;
