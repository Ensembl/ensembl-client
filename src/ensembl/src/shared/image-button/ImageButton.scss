@import 'src/styles/common';

.default {
  position: relative;
  svg {
    width: 100%;
    height: 100%;
    fill: $ens-blue;
    cursor: pointer;
    transition: opacity 0.4s;
    -webkit-transition: opacity 0.4s;
  }
}

.inactive {
  svg {
    fill: $ens-dark-grey;
  }
  svg:hover {
    opacity: 0.8;
  }
}

.active {
  svg {
    fill: $ens-blue;
  }
  svg:hover {
    opacity: 0.8;
  }
}

.disabled {
  svg {
    fill: $ens-grey;
    cursor: not-allowed;
    opacity: 1;
  }
  &:before {
    position: absolute;
    content: '';
    top: 50%;
    border-color: #2c343a;
    border-top: 2px solid red;
    -webkit-transform: rotate(-5deg);
    transform: rotate(-45deg);
    width: 100%;
  }
  &:after {
    position: absolute;
    content: '';
    left: 0px;
    border-color: #2c343a;
    border: 2px solid red;
    border-radius: 50%;
    height: 100%;
    width: 100%;
  }
}
