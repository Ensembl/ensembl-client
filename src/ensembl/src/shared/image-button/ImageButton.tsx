import React from 'react';
import defaultStyles from './ImageButton.scss';
import ImageHolder from './ImageHolder';

export enum ImageButtonStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  DISABLED = 'disabled'
}

type Props = {
  buttonStatus: ImageButtonStatus;
  description: string;
  image: React.FunctionComponent<React.SVGProps<SVGSVGElement>> | string;
  classNames?: any;
  onClick: () => void;
};

const ImageButton = (props: Props) => {
  const buttonProps =
    props.buttonStatus === ImageButtonStatus.DISABLED
      ? {}
      : { onClick: props.onClick };

  return (
    <div {...buttonProps}>
      <ImageHolder
        buttonStatus={props.buttonStatus}
        description={props.description}
        image={props.image}
        classNames={props.classNames ? props.classNames : defaultStyles}
      />
    </div>
  );
};

export default ImageButton;
