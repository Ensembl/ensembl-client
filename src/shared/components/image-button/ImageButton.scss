@import 'src/styles/common';

/* stylelint-disable no-descending-specificity */
// stylelint doesn't like that the 'imageButton button > svg' selector is more specific than ".default svg" and others;
// but moving it below the rest of the selectors feels weird

.imageButton {
  position: relative;
  font-size: 0;

  svg {
    transition: opacity 0.4s;
  }

  button {
    width: 100%;
    height: 100%;
  }

  // if a button's immediate child is an svg, it should be the same size as the button
  button > svg {
    width: 100%;
    height: 100%;
  }
}

.default {
  svg {
    fill: $blue;
  }
}

.selected {
  svg {
    fill: $white;
    background: $blue;
  }
  &svg:hover {
    opacity: 0.8;
  }
}

.unselected {
  svg {
    fill: $blue;
  }
  &svg:hover {
    opacity: 0.8;
  }
}

.disabled {
  svg {
    fill: $grey;
    opacity: 1;
  }
  button {
    cursor: default;
  }
}
