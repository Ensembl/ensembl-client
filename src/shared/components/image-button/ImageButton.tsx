/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { memo } from 'react';
import isEqual from 'lodash/isEqual';
import classNames from 'classnames';

import useHover from 'src/shared/hooks/useHover';

import Tooltip from 'src/shared/components/tooltip/Tooltip';

import { Status } from 'src/shared/types/status';

import imageButtonStyles from './ImageButton.scss';

export type ImageButtonStatus =
  | Status.DEFAULT
  | Status.SELECTED
  | Status.UNSELECTED
  | Status.DISABLED;

export type Props = {
  status: ImageButtonStatus;
  description: string;
  image: React.FunctionComponent<React.SVGProps<SVGSVGElement>> | string;
  className?: string;
  statusClasses?: { [key in ImageButtonStatus]?: string };
  onClick?: () => void;
};

export const ImageButton = (props: Props) => {
  const [hoverRef, isHovered] = useHover<HTMLDivElement>();

  const handleClick = () => {
    props.onClick && props.onClick();
  };

  const buttonProps =
    props.status === Status.DISABLED ? {} : { onClick: handleClick };

  const styles = props.statusClasses
    ? { ...imageButtonStyles, ...props.statusClasses }
    : imageButtonStyles;

  const imageButtonClasses = classNames(
    imageButtonStyles.imageButton,
    props.className,
    styles[props.status]
  );

  const shouldShowTooltip = Boolean(props.description) && isHovered;

  return (
    <div ref={hoverRef} className={imageButtonClasses} {...buttonProps}>
      <button>
        {typeof props.image === 'string' ? (
          <img src={props.image} alt={props.description} />
        ) : (
          <props.image />
        )}
      </button>
      {shouldShowTooltip && (
        <Tooltip anchor={hoverRef.current} autoAdjust={true}>
          {props.description}
        </Tooltip>
      )}
    </div>
  );
};

ImageButton.defaultProps = {
  status: Status.DEFAULT,
  description: ''
};

export default memo(ImageButton, isEqual);
