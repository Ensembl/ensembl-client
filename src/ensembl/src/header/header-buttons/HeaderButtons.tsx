import React, { FunctionComponent } from 'react';
import { connect } from 'react-redux';

import { toggleAccount, toggleLaunchbar } from '../headerActions';

import ImageButton, {
  ImageButtonStatus
} from 'src/shared/image-button/ImageButton';

import { ReactComponent as LaunchbarIcon } from 'static/img/header/launchbar.svg';
import { ReactComponent as UserIcon } from 'static/img/header/user-grey.svg';

import styles from './HeaderButtons.scss';

type StateProps = {};

type DispatchProps = {
  toggleLaunchbar: () => void;
  toggleAccount: () => void;
};

type OwnProps = {};

type HeaderButtonsProps = StateProps & DispatchProps & OwnProps;

export const HeaderButtons: FunctionComponent<HeaderButtonsProps> = (props) => (
  <div className={styles.headerButtons}>
    <div className={styles.buttonWrapper}>
      <ImageButton
        image={LaunchbarIcon}
        description="Ensembl app launchbar"
        onClick={props.toggleLaunchbar}
      />
    </div>
    <div className={styles.buttonWrapper}>
      <ImageButton
        image={UserIcon}
        description="Ensembl account"
        buttonStatus={ImageButtonStatus.DISABLED}
        classNames={{
          [ImageButtonStatus.DISABLED]: styles.headerButtonDisabled
        }}
      />
    </div>
  </div>
);

const mapStateToProps = (): StateProps => ({});

const mapDispatchToProps: DispatchProps = {
  toggleAccount,
  toggleLaunchbar
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HeaderButtons);
