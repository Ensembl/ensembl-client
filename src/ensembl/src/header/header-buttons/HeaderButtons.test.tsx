import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { HeaderButtons } from './HeaderButtons';
import ImageButton, {
  ImageButtonStatus
} from 'src/shared/components/image-button/ImageButton';

describe('<HeaderButtons />', () => {
  let toggleLaunchbarFn: () => void;
  let toggleAccountFn: () => void;
  let shallowWrapper: ShallowWrapper;

  beforeEach(() => {
    toggleLaunchbarFn = jest.fn();
    toggleAccountFn = jest.fn();
    shallowWrapper = shallow(
      <HeaderButtons
        toggleAccount={toggleAccountFn}
        toggleLaunchbar={toggleLaunchbarFn}
      />
    );
  });

  test('contains button for toggling launchbar', () => {
    const launchbarButton = shallowWrapper
      .find(ImageButton)
      .filterWhere(
        (wrapper) => wrapper.prop('description') === 'Ensembl app launchbar'
      );
    expect(launchbarButton.prop('onClick')).toBe(toggleLaunchbarFn);
  });

  test('contains disabled user account button', () => {
    const launchbarButton = shallowWrapper
      .find(ImageButton)
      .filterWhere(
        (wrapper) => wrapper.prop('description') === 'Ensembl account'
      );
    expect(launchbarButton.prop('buttonStatus')).toBe(
      ImageButtonStatus.DISABLED
    );
  });
});
