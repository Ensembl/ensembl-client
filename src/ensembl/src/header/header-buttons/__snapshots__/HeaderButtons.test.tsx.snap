// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderButtons /> renders correctly 1`] = `
<div
  className="headerButtons"
>
  <button
    className="launchbarButton"
    onClick={[MockFunction]}
  >
    <img
      alt="toggle launchbar"
      src={Object {}}
      title="Launchbar"
    />
  </button>
  <button
    className="accountButton"
    onClick={[MockFunction]}
  >
    <img
      alt="toggle account"
      src={Object {}}
      title="Account"
    />
  </button>
</div>
`;
